docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,402 @@
+members.
+
+.. code-block:: python
+
+    (a, b, c) : Tuple<A, B, C>
+
+    (add(add(a, b), c), d) : Tuple<Tensor<float32, (10, 10)>, Tensor<float32, (100, 100)>>
+
+See :py:class:`~tvm.relay.expr.Tuple` for its definition and documentation.
+
+Tuple Projection
+~~~~~~~~~~~~~~~~
+
+A tuple must be indexed by an integer constant in order to extract a
+particular member of the tuple. Projections are 0-indexed.
+
+For example, the below projection evaluates to :code:`b`:
+
+.. code-block:: python
+
+   (a, b, c).1
+
+See :py:class:`~tvm.relay.expr.TupleGetItem` for its definition and documentation.
+
+Function Expressions
+~~~~~~~~~~~~~~~~~~~~
+
+Functions are first class in Relay and can be used in any expression
+position. Function expressions are the same as global functions and
+use nearly the same syntax (besides not including a name), but do
+not have a globally unique name.
+
+.. code-block:: python
+
+    fun (%x : Tensor<float32, (10, 10)>, y: Tensor<float32, (10, 10)>
+                -> Tensor<float32, (10, 10)> { add(%x, %y) }
+
+Note that function expressions evaluate to closure values. Closures
+store values for all free variables in their body. A free variable
+is a variable defined outside the scope of the function's body,
+which means that if a function expression references a local variable
+in the outer scope that is not shadowed by the parameters,
+the closure will store the value for that local variable and
+use that value when the function is called, even if original
+local variable has gone out of scope.
+
+For example, in the below example, :code:`%z` will evaluate to a tensor
+of zero values because the closure for :code:`%f` stores the value of
+:code:`%x` at the pointer where :code:`%f` was defined.
+
+.. code-block::
+
+    let %g = fun () {
+      let %x = Constant(0, float32, (10, 10));
+      # x is a free variable in the below function
+      fun (%y) { multiply(%y, %x) }
+    };
+    # the %x in %g's body is not in scope anymore
+    # %f is a closure where %x maps to Constant(0, float32, (10, 10))
+    let %f = %g();
+    let %x = Constant(1, float32, (10, 10));
+    let %z = %f(%x)
+
+A recursive function expression can be defined using a `let` binding,
+as here:
+
+.. code-block:: python
+
+    let %fact = fun (%x : Tensor<float32, (10, 10)>) -> Tensor<float32, (10, 10)> {
+        if (equal(%x, Constant(1, float32, (10, 10))) {
+            Constant(0, float32, (10, 10))
+        } else {
+            multiply(%x,  %fact(subtract(%x, Constant(1, float32, (10, 10))))
+        }
+    };
+    %fact(10)
+
+See :py:class:`~tvm.relay.expr.Function` for its definition and documentation.
+
+Let Binding
+~~~~~~~~~~~
+
+A `let` binding is an immutable local variable binding,
+allowing the user to bind an expression to a name.
+
+A `let` binding contains a local variable,
+an optional type annotation, a value, and a body expression
```

Comment:
We should probably comment on explicit vs. implicit sharing here @MarisaKirisame